### ISSUE TITLE
Problem installing package from custom channel with same name as official package

### ISSUE THREAD

**Issue by lschr** [conda/conda#2804]
With conda 4.1* (<4.1 works perfectly) there are problems if I have in one of my custom channels a package of the same name (although different version) as in the official repository.

I will demonstrate the problem using `libffi`as an example.

- No custom channel: Everything works as expected
```
$ conda search libffi
Fetching package metadata .......
libffi                       3.0.13                        0  defaults        
                             3.2.1                         0  defaults
$ conda install libffi
Fetching package metadata .......
Solving package specifications: ..........

Package plan for installation in environment <env>:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    libffi-3.2.1               |                0          36 KB

The following NEW packages will be INSTALLED:

    libffi: 3.2.1-0

Proceed ([y]/n)?
```

- Channel with a `file://` URL: Apparently the package from the custom channel is preferred, although it has a lower version number.
```
$ conda search libffi
Fetching package metadata .........
libffi                       3.0.13                        0  defaults        
                             3.0.13                        4  file://<path>/linux-64
                             3.2.1                         0  defaults
$ conda install libffi
Fetching package metadata .........
Solving package specifications: ..........

Package plan for installation in environment <env>:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    libffi-3.0.13              |                4          38 KB  file://<path>/linux-64

The following NEW packages will be INSTALLED:

    libffi: 3.0.13-4 file://<path>

Proceed ([y]/n)?
```
- Channel with a `https://` URL (and basic http auth): Conda says that the package is already installed although it isn't.
```
$ conda search libffi
Fetching package metadata: ......
libffi                       3.0.13                        0  defaults        
                             3.0.13                        4  https://<user>:<pass>@<url>/linux-64/
                             3.2.1                         0  defaults
$ conda install libffi
Fetching package metadata .........
Solving package specifications: ..........

# All requested packages already installed.
# packages in environment at <env>:
#

```

I tried the same things with conda 4.0.9 and in every case it will install libffi 3.2.1 as expected.

**Comment by mcg1969** [comment]
Conda 4.1 now prioritizes channels by the order in which they are offered in `.condarc`. That means that it will indeed prefer an _older_ version of a package if it is found in a _higher-priority_ channel.

If you would like to disable this behavior and return roughly to the 4.0.x behavior, you can use `channel_priority: False` in your `.condarc`.

It's not clear to me what's going on in the third case. Can you go ahead and reproduce this one for me, and then offer both the output of `conda list` and `conda info`? Thanks.

**Comment by lschr** [comment]
Thanks for the tip about `channel_priority`.

I am not quite sure what you mean by "reproducing" the issue… It is reproducible with all packages I have in both the custom and the default channels.

Anyways, I created a new env and tried to `conda install libffi`, which again told me that it was already installed.
```
$ conda list
# packages in environment at /home/lukas/Software/anaconda3/envs/test_channel:
#
openssl                   1.0.2h                        1  
pip                       8.1.2                    py35_0  
python                    3.5.1                         0  
readline                  6.2                           2  
setuptools                23.0.0                   py35_0  
sqlite                    3.13.0                        0  
tk                        8.5.18                        0  
wheel                     0.29.0                   py35_0  
xz                        5.0.5                         1  
zlib                      1.2.8                         3
$ conda info
Current conda install:

             platform : linux-64
        conda version : 4.1.2
  conda-build version : 1.21.1
       python version : 3.5.1.final.0
     requests version : 2.10.0
     root environment : /home/lukas/Software/anaconda3  (writable)
  default environment : /home/lukas/Software/anaconda3/envs/test_channel
     envs directories : /home/lukas/Software/anaconda3/envs
        package cache : /home/lukas/Software/anaconda3/pkgs
         channel URLs : https://<user>:<password>@<url>/linux-64/
                        https://<user>:<password>@<url>/noarch/
                        https://repo.continuum.io/pkgs/free/linux-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/linux-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : /home/lukas/.condarc
    is foreign system : False
```

Apparently, when trying to install such a package that is available in both channels, its dependencies get installed, but not the package itself. After trying to install spyder (which is in both channels) I get
```
$ conda list
# packages in environment at /home/lukas/Software/anaconda3/envs/test_channel:
#
alabaster                 0.7.8                    py35_0  
astroid                   1.4.4                    py35_0  
babel                     2.3.3                    py35_0  
backports                 1.0                      py35_0  
decorator                 4.0.10                   py35_0  
docutils                  0.12                     py35_2  
entrypoints               0.2.2                    py35_0  
fontconfig                2.11.1                        5  
freetype                  2.5.5                         1  
get_terminal_size         1.0.0                    py35_0  
imagesize                 0.7.1                    py35_0  
ipykernel                 4.3.1                    py35_0  
ipython                   4.2.0                    py35_0  
ipython_genutils          0.1.0                    py35_0  
jinja2                    2.8                      py35_1  
jsonschema                2.5.1                    py35_0  
jupyter_client            4.3.0                    py35_0  
jupyter_core              4.1.0                    py35_0  
lazy-object-proxy         1.2.1                    py35_0  
libpng                    1.6.17                        0  
libsodium                 1.0.10                        0  
libxml2                   2.9.2                         0  
logilab-common            1.0.2                    py35_0  
markupsafe                0.23                     py35_2  
mistune                   0.7.2                    py35_0  
nbconvert                 4.2.0                    py35_0  
nbformat                  4.0.1                    py35_0  
openssl                   1.0.2h                        1  
path.py                   8.2.1                    py35_0  
pep8                      1.7.0                    py35_0  
pexpect                   4.0.1                    py35_0  
pickleshare               0.7.2                    py35_0  
pip                       8.1.2                    py35_0  
psutil                    4.3.0                    py35_0  
ptyprocess                0.5.1                    py35_0  
pyflakes                  1.1.0                    py35_0  
pygments                  2.1.3                    py35_0  
pylint                    1.5.4                    py35_0  
pyqt                      4.11.4                   py35_3  
python                    3.5.1                         0  
pytz                      2016.4                   py35_0  
pyzmq                     15.2.0                   py35_1  
qt                        4.8.7                         3  
qtawesome                 0.3.3                    py35_0  
qtconsole                 4.2.1                    py35_0  
qtpy                      1.0.2                    py35_0  
readline                  6.2                           2  
rope                      0.9.4                    py35_1  
setuptools                23.0.0                   py35_0  
simplegeneric             0.8.1                    py35_1  
sip                       4.16.9                   py35_0  
six                       1.10.0                   py35_0  
snowballstemmer           1.2.1                    py35_0  
sphinx                    1.4.1                    py35_0  
sphinx_rtd_theme          0.1.9                    py35_0  
sqlite                    3.13.0                        0  
tk                        8.5.18                        0  
tornado                   4.3                      py35_1  
traitlets                 4.2.1                    py35_0  
wheel                     0.29.0                   py35_0  
wrapt                     1.10.6                   py35_0  
xz                        5.0.5                         1  
zeromq                    4.1.4                         0  
zlib                      1.2.8                         3
```
i. e. spyder is missing.

**Comment by kalefranz** [comment]
@lschr 

> Anyways, I created a new env and tried to `conda install libffi`, which again told me that it was already installed.

You should show us the shell line that include `conda install libffi` and all of its output.  "told me that it was already installed" can actually mean several different things, and the full output helps us diagnose what's going on.

> After trying to install spyder (which is in both channels) I get

Again you didn't include the full input/output of the session.  Don't worry about the report being too long.  We'd rather have too much information than too little.  Right now, there's not enough information for us to act on here.

**Comment by lschr** [comment]
@kalefranz My original report already contains the output for `conda search libffi` and `conda install libffi`. In the first two cases, which seem two work as expected, I aborted after conda asked me whether to proceed installing, but in the third (i. e. the interesting) case this is indeed the full output.

**Comment by lschr** [comment]
@kalefranz Here is also the output of
- creating a new env
- activating it
- trying to install spyder (which exists also in my https channel in a newer version)
```
$ conda create -n test_channel_2 python=3.5
Fetching package metadata .........
Solving package specifications: ..........

Package plan for installation in environment /home/lukas/Software/anaconda3/envs/test_channel_2:

The following NEW packages will be INSTALLED:

    openssl:    1.0.2h-1     
    pip:        8.1.2-py35_0 
    python:     3.5.1-5      
    readline:   6.2-2        
    setuptools: 23.0.0-py35_0
    sqlite:     3.13.0-0     
    tk:         8.5.18-0     
    wheel:      0.29.0-py35_0
    xz:         5.2.2-0      
    zlib:       1.2.8-3      

Proceed ([y]/n)? 

Linking packages ...
[      COMPLETE      ]|##############################################################################| 100%
#
# To activate this environment, use:
# $ source activate test_channel_2
#
# To deactivate this environment, use:
# $ source deactivate
#
$ source activate test_channel_2                  
prepending /home/lukas/Software/anaconda3/envs/test_channel_2/bin to PATH
$ conda install spyder
Fetching package metadata .........
Solving package specifications: ..........

Package plan for installation in environment /home/lukas/Software/anaconda3/envs/test_channel_2:

The following NEW packages will be INSTALLED:

    alabaster:         0.7.8-py35_0 
    astroid:           1.4.4-py35_0 
    babel:             2.3.3-py35_0 
    backports:         1.0-py35_0   
    decorator:         4.0.10-py35_0
    docutils:          0.12-py35_2  
    entrypoints:       0.2.2-py35_0 
    fontconfig:        2.11.1-6     
    freetype:          2.5.5-1      
    get_terminal_size: 1.0.0-py35_0 
    imagesize:         0.7.1-py35_0 
    ipykernel:         4.3.1-py35_0 
    ipython:           4.2.0-py35_0 
    ipython_genutils:  0.1.0-py35_0 
    jinja2:            2.8-py35_1   
    jsonschema:        2.5.1-py35_0 
    jupyter_client:    4.3.0-py35_0 
    jupyter_core:      4.1.0-py35_0 
    lazy-object-proxy: 1.2.1-py35_0 
    libpng:            1.6.22-0     
    libsodium:         1.0.10-0     
    libxml2:           2.9.2-0      
    logilab-common:    1.0.2-py35_0 
    markupsafe:        0.23-py35_2  
    mistune:           0.7.2-py35_0 
    nbconvert:         4.2.0-py35_0 
    nbformat:          4.0.1-py35_0 
    path.py:           8.2.1-py35_0 
    pep8:              1.7.0-py35_0 
    pexpect:           4.0.1-py35_0 
    pickleshare:       0.7.2-py35_0 
    psutil:            4.3.0-py35_0 
    ptyprocess:        0.5.1-py35_0 
    pyflakes:          1.2.3-py35_0 
    pygments:          2.1.3-py35_0 
    pylint:            1.5.4-py35_0 
    pyqt:              4.11.4-py35_3
    pytz:              2016.4-py35_0
    pyzmq:             15.2.0-py35_1
    qt:                4.8.7-3      
    qtawesome:         0.3.3-py35_0 
    qtconsole:         4.2.1-py35_0 
    qtpy:              1.0.2-py35_0 
    rope:              0.9.4-py35_1 
    simplegeneric:     0.8.1-py35_1 
    sip:               4.16.9-py35_0
    six:               1.10.0-py35_0
    snowballstemmer:   1.2.1-py35_0 
    sphinx:            1.4.1-py35_0 
    sphinx_rtd_theme:  0.1.9-py35_0 
    tornado:           4.3-py35_1   
    traitlets:         4.2.1-py35_0 
    wrapt:             1.10.6-py35_0
    zeromq:            4.1.4-0      

Proceed ([y]/n)? 

Linking packages ...
[      COMPLETE      ]|##############################################################################| 100%
```

**Comment by lschr** [comment]
The issue is still present in conda 4.1.11 (while still absent in conda 4.0.11). It is also reproducible on Windows machines.

Is there anything I can do to help fix the problem?

**Comment by elehcim** [comment]
While investigating on #3459, I found that the behaviour you are reporting here (I'm also affected) is due to the fact that your repo is behind basic http auth. In particular the representation of your package just before resolving the list of the dependent packages is something like:

    `https://<user>:<password>@<url>::spyder-<version>`

In [conda/resolve.py:983](https://github.com/conda/conda/blob/e038eca300152e75d7d213a091cee95daf36a922/conda/resolve.py#L983) packages with a "@" in the name are skipped. In fact the list of the packages to be installed computed [here](https://github.com/conda/conda/blob/e038eca300152e75d7d213a091cee95daf36a922/conda/resolve.py#L978) is something like:

    ['sqlite-3.13.0-0.tar.bz2', '@s@sqlite', 'cairo-1.12.18-6.tar.bz2', '@s@cairo', 'libpng-1.6.22-0.tar.bz2', '@s@libpng', ...]

For this your package is cleaned and removed from the list of the packages to be installed.

As a workaround (but I don't know how much robust) I changed the [clean function](https://github.com/conda/conda/blob/e038eca300152e75d7d213a091cee95daf36a922/conda/resolve.py#L981) in this way:

    def clean(sol):
        return [q for q in (C.from_index(s) for s in sol)
                if q and q[0] != '!' and not (('@' in q) ^ ('::' in q))]

meaning that the package shouldn't be remove if both `'@'` and `'::'` substrings are in the package name.
I'm exploiting the fact that the name of the package coming from custom channels has `'::'` as separator between the channel and the package name. Probably there is a better way.

I also think that the most robust way to deal with http basic auth is to remove all the authentication parts from the urls and use them only when performing `requests.get` in [`fetch_repodata`](https://github.com/conda/conda/blob/e038eca300152e75d7d213a091cee95daf36a922/conda/fetch.py#L113) and [`fetch_pkg`](https://github.com/conda/conda/blob/e038eca300152e75d7d213a091cee95daf36a922/conda/fetch.py#L341), for example overwriting `CondaSession.auth` member. But this is related to #3459. Maybe I can propose a PR soon.

**Comment by kalefranz** [comment]
addressed in #3544 via #3550
